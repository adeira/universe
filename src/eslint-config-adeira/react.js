// @flow

const { react } = require('./presetsConfig');
const changeNextVersionErrorLevel = require('./changeNextVersionErrorLevel');
const getCommonConfig = require('./getCommonConfig');

/*::

import type { EslintConfig } from './EslintConfig.flow';

*/

const WARN = 1;

const preset = ({
  ...getCommonConfig(changeNextVersionErrorLevel(react.rules, WARN)),
  plugins: react.plugins,
} /*: EslintConfig */);

module.exports = preset;
