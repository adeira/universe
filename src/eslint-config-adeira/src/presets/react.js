// @flow

const { ERROR, OFF, WARN } = require('../constants');
const detectReactVersion = require('../detectReactVersion');

/*::

import type { EslintConfig } from '../EslintConfig.flow';

*/

module.exports = ({
  plugins: [
    'eslint-plugin-react',
    'eslint-plugin-react-hooks',
    'eslint-plugin-react-native',
    'eslint-plugin-jsx-a11y',
  ],
  rules: {
    // React (https://github.com/yannickcr/eslint-plugin-react)
    'react/boolean-prop-naming': OFF,
    'react/button-has-type': [ERROR, { button: true, submit: true, reset: false }],
    'react/default-props-match-prop-types': OFF,
    'react/destructuring-assignment': OFF,
    'react/display-name': OFF,
    'react/forbid-component-props': OFF,
    'react/forbid-dom-props': [
      ERROR,
      {
        forbid: [
          'class', // https://reactjs.org/docs/dom-elements.html#classname
          'for', // https://reactjs.org/docs/dom-elements.html#htmlfor
        ],
      },
    ],
    'react/forbid-elements': OFF,
    'react/forbid-foreign-prop-types': OFF,
    'react/forbid-prop-types': OFF,
    'react/jsx-boolean-value': OFF,
    'react/jsx-curly-brace-presence': OFF,
    'react/jsx-filename-extension': OFF,
    'react/jsx-fragments': OFF,
    'react/jsx-handler-names': OFF,
    'react/jsx-key': ERROR,
    'react/jsx-max-depth': OFF,
    'react/jsx-no-bind': [
      ERROR,
      {
        ignoreDOMComponents: true,
        ignoreRefs: true,
        allowArrowFunctions: true,
        allowFunctions: false,
        allowBind: false,
      },
    ],
    'react/jsx-no-comment-textnodes': ERROR,
    'react/jsx-no-constructed-context-values': ERROR,
    'react/jsx-no-duplicate-props': ERROR,
    'react/jsx-no-literals': OFF,
    'react/jsx-no-target-blank': ERROR,
    'react/jsx-no-undef': ERROR,
    'react/jsx-no-useless-fragment': ERROR,
    'react/jsx-pascal-case': ERROR,
    'react/jsx-props-no-spreading': OFF,
    'react/jsx-sort-default-props': OFF,
    'react/jsx-sort-props': OFF,
    'react/jsx-uses-react': WARN,
    'react/jsx-uses-vars': ERROR,
    'react/no-access-state-in-setstate': ERROR,
    'react/no-array-index-key': OFF,
    'react/no-children-prop': OFF,
    'react/no-danger': OFF,
    'react/no-danger-with-children': ERROR,
    'react/no-deprecated': ERROR,
    'react/no-did-mount-set-state': OFF, // https://github.com/airbnb/javascript/issues/684#issuecomment-264094930
    'react/no-did-update-set-state': ERROR,
    'react/no-direct-mutation-state': ERROR,
    'react/no-find-dom-node': ERROR,
    'react/no-is-mounted': ERROR,
    'react/no-multi-comp': [ERROR, { ignoreStateless: true }],
    'react/no-redundant-should-component-update': ERROR,
    'react/no-render-return-value': ERROR,
    'react/no-set-state': OFF,
    'react/no-string-refs': OFF,
    'react/no-this-in-sfc': ERROR,
    'react/no-typos': OFF,
    'react/no-unescaped-entities': ERROR,
    'react/no-unknown-property': ERROR,
    'react/no-unsafe': OFF,
    'react/no-unstable-nested-components': [ERROR, { allowAsProps: true }],
    'react/no-unused-prop-types': OFF,
    'react/no-unused-state': OFF, // Enable when they fix this issue: https://github.com/yannickcr/eslint-plugin-react/issues/1910
    'react/no-will-update-set-state': OFF,
    'react/prefer-es6-class': [ERROR, 'always'],
    'react/prefer-exact-props': OFF, // see: fb-flow/use-exact-by-default-object-type
    'react/prefer-read-only-props': OFF,
    'react/prefer-stateless-function': OFF,
    'react/prop-types': OFF, // we use Flow instead,
    'react/react-in-jsx-scope': OFF,
    'react/require-default-props': OFF,
    'react/require-optimization': OFF,
    'react/require-render-return': ERROR,
    'react/self-closing-comp': ERROR,
    'react/sort-comp': [
      ERROR,
      {
        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-comp.md
        order: [
          'type-annotations',
          'static-variables',
          'static-methods',
          'instance-variables',
          'lifecycle',
          'everything-else',
          'rendering',
        ],
        groups: {
          lifecycle: [
            'displayName',
            'propTypes',
            'contextTypes',
            'childContextTypes',
            'mixins',
            'statics',
            'defaultProps',
            'constructor',
            'getDefaultProps',
            'state',
            'getInitialState',
            'getChildContext',
            'getDerivedStateFromProps', // this is added
            'getDerivedStateFromError', // this is added
            'componentWillMount',
            'UNSAFE_componentWillMount',
            'componentDidMount',
            'componentWillReceiveProps',
            'UNSAFE_componentWillReceiveProps',
            'shouldComponentUpdate',
            'componentWillUpdate',
            'UNSAFE_componentWillUpdate',
            'getSnapshotBeforeUpdate',
            'componentDidUpdate',
            'componentDidCatch',
            'componentWillUnmount',
          ],
          rendering: ['/^render.+$/', 'render'],
        },
      },
    ],
    'react/sort-prop-types': OFF,
    'react/state-in-constructor': OFF,
    'react/static-property-placement': OFF,
    'react/style-prop-object': ERROR,
    'react/void-dom-elements-no-children': ERROR,
    'react/function-component-definition': OFF,
    'react/jsx-no-script-url': ERROR,
    'react/no-adjacent-inline-elements': OFF,

    // React Hooks (https://www.npmjs.com/package/eslint-plugin-react-hooks)
    'react-hooks/rules-of-hooks': ERROR,
    'react-hooks/exhaustive-deps': [
      ERROR,
      {
        // https://recoiljs.org/docs/introduction/installation/#eslint
        additionalHooks: 'useRecoilCallback', // should be regex!
      },
    ],

    // React Native (https://github.com/intellicode/eslint-plugin-react-native)
    'react-native/no-color-literals': OFF,
    'react-native/no-inline-styles': OFF,
    'react-native/no-raw-text': OFF, // TODO: takes normal React into account as well, not only RN
    'react-native/no-single-element-style-arrays': ERROR,
    'react-native/no-unused-styles': ERROR,
    'react-native/sort-styles': OFF,
    'react-native/split-platform-components': OFF,

    // React Accessibility (https://github.com/evcohen/eslint-plugin-jsx-a11y)
    'jsx-a11y/accessible-emoji': ERROR,
    'jsx-a11y/alt-text': ERROR,
    'jsx-a11y/anchor-has-content': ERROR,
    'jsx-a11y/anchor-is-valid': ERROR,
    'jsx-a11y/aria-activedescendant-has-tabindex': OFF,
    'jsx-a11y/aria-props': ERROR,
    'jsx-a11y/aria-proptypes': ERROR,
    'jsx-a11y/aria-role': ERROR,
    'jsx-a11y/aria-unsupported-elements': ERROR,
    'jsx-a11y/autocomplete-valid': ERROR,
    'jsx-a11y/click-events-have-key-events': OFF,
    'jsx-a11y/control-has-associated-label': OFF,
    'jsx-a11y/heading-has-content': ERROR,
    'jsx-a11y/html-has-lang': ERROR,
    'jsx-a11y/iframe-has-title': ERROR,
    'jsx-a11y/img-redundant-alt': OFF,
    'jsx-a11y/interactive-supports-focus': ERROR,
    'jsx-a11y/label-has-associated-control': OFF,
    'jsx-a11y/label-has-for': OFF,
    'jsx-a11y/lang': ERROR,
    'jsx-a11y/media-has-caption': OFF,
    'jsx-a11y/mouse-events-have-key-events': OFF,
    'jsx-a11y/no-access-key': ERROR,
    'jsx-a11y/no-autofocus': [ERROR, { ignoreNonDOM: true }],
    'jsx-a11y/no-distracting-elements': [ERROR, { elements: ['marquee', 'blink'] }],
    'jsx-a11y/no-interactive-element-to-noninteractive-role': OFF,
    'jsx-a11y/no-noninteractive-element-interactions': OFF,
    'jsx-a11y/no-noninteractive-element-to-interactive-role': OFF,
    'jsx-a11y/no-noninteractive-tabindex': ERROR,
    'jsx-a11y/no-onchange': OFF,
    'jsx-a11y/no-redundant-roles': ERROR,
    'jsx-a11y/no-static-element-interactions': OFF,
    'jsx-a11y/role-has-required-aria-props': ERROR,
    'jsx-a11y/role-supports-aria-props': ERROR,
    'jsx-a11y/scope': ERROR,
    'jsx-a11y/tabindex-no-positive': ERROR,
  },
  settings: {
    react: {
      version: detectReactVersion(),
    },
  },
} /*: EslintConfig */);
