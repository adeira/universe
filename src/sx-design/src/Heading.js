// @flow

import * as React from 'react';
import sx from '@adeira/sx';

import HeadingLevel from './HeadingLevel';

type Props = {|
  +children: React.Node,
  +xstyle?: $FlowFixMe, // TODO: https://github.com/adeira/universe/issues/1584
|};

// https://web.dev/headings-and-landmarks/
// https://github.com/jonathantneal/h-element-spec/issues/1
export default function Heading(props: Props): React.Node {
  const removeMe = new Map(); // TODO: https://github.com/adeira/universe/pull/1583
  const className = sx(props.xstyle ?? removeMe);
  return (
    <HeadingLevel.Consumer>
      {(level) => {
        const Heading = `h${Math.min(level, 6)}`;
        return (
          <Heading className={className === '' ? undefined : className}>{props.children}</Heading>
        );
      }}
    </HeadingLevel.Consumer>
  );
}
