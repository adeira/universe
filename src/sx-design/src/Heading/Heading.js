// @flow

import * as React from 'react';
import sx, { type AllCSSProperties } from '@adeira/sx';

import HeadingLevel from '../HeadingLevel';

type Props = {|
  +children: React.Node,
  +xstyle?: AllCSSProperties,
|};

/**
 * Heading component automatically renders the right `h[1-6]` HTML element based on the location in
 * the React tree. To advance the heading level you should nest it inside `<Section />` component.
 *
 * For more info please visit:
 *
 * - https://web.dev/headings-and-landmarks/
 * - https://github.com/jonathantneal/h-element-spec/issues/1
 */
export default function Heading(props: Props): React.Node {
  return (
    <HeadingLevel.Consumer>
      {(level) => {
        const Heading = `h${Math.min(level, 6)}`;
        return <Heading className={sx(styles.heading, props.xstyle)}>{props.children}</Heading>;
      }}
    </HeadingLevel.Consumer>
  );
}

const styles = sx.create({
  heading: {
    color: 'rgba(var(--sx-text-color))',
  },
});
