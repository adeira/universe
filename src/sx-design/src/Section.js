// @flow

import * as React from 'react';
import sx, { type AllCSSProperties } from '@adeira/sx';

import HeadingLevel from './HeadingLevel';

type Props = {|
  +children: React.Node,
  +xstyle?: AllCSSProperties,
|};

export default function Section(props: Props): React.Node {
  return (
    <HeadingLevel.Consumer>
      {(level) => (
        <HeadingLevel.Provider value={level + 1}>
          {/* eslint-disable-next-line react/forbid-elements */}
          <section className={sx(props.xstyle)}>{props.children}</section>
        </HeadingLevel.Provider>
      )}
    </HeadingLevel.Consumer>
  );
}
