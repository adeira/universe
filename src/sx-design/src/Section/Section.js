// @flow

import * as React from 'react';
import sx, { type AllCSSProperties } from '@adeira/sx';

import HeadingLevel from '../HeadingLevel';

type Props = {
  +children: React.Node,
  +xstyle?: AllCSSProperties,
};

/**
 * Section component is used to automatically advance Heading component level.
 *
 * @deprecated Use <Text as="â€¦" /> instead. It makes the code more predictable and doesn't affect
 * other seemingly unrelated components (like <Header/>).
 */
export default function Section(props: Props): React.Node {
  return (
    <HeadingLevel.Consumer>
      {(level) => (
        <HeadingLevel.Provider value={level + 1}>
          <section className={sx(styles.section, props.xstyle)}>{props.children}</section>
        </HeadingLevel.Provider>
      )}
    </HeadingLevel.Consumer>
  );
}

const styles = sx.create({
  section: {
    color: 'rgba(var(--sx-foreground))',
  },
});
