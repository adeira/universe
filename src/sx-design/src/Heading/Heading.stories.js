/**
 * https://storybook.js.org/docs/react/writing-stories/introduction
 * @flow
 */

import React from 'react';
import sx from '@adeira/sx';

import Heading from './Heading';
import Section from '../Section/Section';
import type { StoryTemplate } from '../types';

// ðŸ‘‡ This default export determines where your story goes in the story list
export default {
  title: 'Components/Heading',
  component: Heading,
  argTypes: {
    // hide children property from the Storybook
    children: {
      table: {
        disable: true,
      },
    },
  },
};

// ðŸ‘‡ We create a "template" of how args map to rendering
const Template = (args) => (
  <>
    <Heading {...args}>Top level heading (h1)</Heading>
    <Section>
      <Heading {...args}>
        Heading inside <code>{'<Section />'}</code> component (h2)
      </Heading>
      <Section>
        <Heading {...args}>
          Heading inside another <code>{'<Section />'}</code> component (h3)
        </Heading>
        <Section>
          <Heading {...args}>
            Heading inside another <code>{'<Section />'}</code> component (h4)
          </Heading>
          <Section>
            <Heading {...args}>
              Heading inside another <code>{'<Section />'}</code> component (h5)
            </Heading>
            <Section>
              <Heading {...args}>
                Heading inside another <code>{'<Section />'}</code> component (h6)
              </Heading>
              <Section>
                <Heading {...args}>
                  Heading inside another <code>{'<Section />'}</code> component (still h6)
                </Heading>
              </Section>
            </Section>
          </Section>
        </Section>
      </Section>
    </Section>
  </>
);

/* eslint-disable sx/no-unused-stylesheet */
const styles = sx.create({
  default: {
    color: 'blue',
  },
});
/* eslint-enable sx/no-unused-stylesheet */

// ðŸ‘‡ Each story then reuses that template
export const BasicHeading: StoryTemplate<typeof Heading> = Template.bind({});
BasicHeading.storyName = 'Basic';

export const CustomHeading: StoryTemplate<typeof Heading> = Template.bind({});
CustomHeading.storyName = 'Custom style';
CustomHeading.args = {
  xstyle: styles.default,
};
