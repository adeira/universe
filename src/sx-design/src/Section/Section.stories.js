/**
 * https://storybook.js.org/docs/react/writing-stories/introduction
 * @flow
 */

import sx from '@adeira/sx';
import React from 'react';

import Section from './Section';
import type { StoryTemplate } from '../types';

// 👇 This default export determines where your story goes in the story list
export default {
  title: 'Example/Section',
  component: Section,
  argTypes: {
    // hide children property from the Storybook
    children: {
      table: {
        disable: true,
      },
    },
  },
};

// 👇 We create a "template" of how args map to rendering
const Template = (args) => (
  <Section {...args}>
    Not much to see here (check the <strong>Docs</strong>{' '}
    <span role="img" aria-label="index finger pointing up">
      ☝️
    </span>
    )
  </Section>
);

/* eslint-disable sx/no-unused-stylesheet */
const styles = sx.create({
  default: {
    color: 'blue',
  },
});
/* eslint-enable sx/no-unused-stylesheet */

// 👇 Each story then reuses that template
export const BasicSection: StoryTemplate<typeof Section> = Template.bind({});
BasicSection.storyName = 'Basic';

export const CustomSection: StoryTemplate<typeof Section> = Template.bind({});
CustomSection.storyName = 'Custom style';
CustomSection.args = {
  xstyle: styles.default,
};
