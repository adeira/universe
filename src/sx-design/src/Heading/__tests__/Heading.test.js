// @flow

import * as React from 'react';
import { render } from '@testing-library/react';
import prettier from 'prettier';

import Heading from '../Heading';
import Section from '../../Section/Section';

it('renders top level by default', () => {
  const { container } = render(<Heading>level one</Heading>);
  expect(container.innerHTML).toMatchInlineSnapshot(`"<h1 class=\\"_1cJrQ0\\">level one</h1>"`);
});

it('renders H1 tags in parallel', () => {
  const { container } = render(
    <>
      <Heading>level one</Heading>
      <Heading>level one</Heading>
    </>,
  );
  expect(container.innerHTML).toMatchInlineSnapshot(
    `"<h1 class=\\"_1cJrQ0\\">level one</h1><h1 class=\\"_1cJrQ0\\">level one</h1>"`,
  );
});

// <Heading>
//   <Heading>level one</Heading>
// </Heading>
it.todo("doesn't allow nested headers");

it('renders maximum 6 levels', () => {
  const { container } = render(
    <>
      <Heading>level one</Heading>
      <Section>
        <Heading>level two</Heading>
        <Section>
          <Heading>level three</Heading>
          <Section>
            <Heading>level four</Heading>
            <Section>
              <Heading>level five</Heading>
              <Section>
                <Heading>level six</Heading>
                <Section>
                  <Heading>level six again</Heading>
                  <Heading>level six again</Heading>
                </Section>
                <Heading>level six</Heading>
              </Section>
              <Heading>level five</Heading>
            </Section>
            <Heading>level four</Heading>
          </Section>
          <Heading>level three</Heading>
        </Section>
        <Heading>level two</Heading>
      </Section>
    </>,
  );

  expect(
    prettier.format(container.innerHTML, {
      filepath: 'test.html',
    }),
  ).toMatchSnapshot();
});
