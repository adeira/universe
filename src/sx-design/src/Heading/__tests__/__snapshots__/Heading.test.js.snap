// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders maximum 6 levels 1`] = `
"<h1 class=\\"_1cJrQ0\\">level one</h1>
<section class=\\"_1cJrQ0\\">
  <h2 class=\\"_1cJrQ0\\">level two</h2>
  <section class=\\"_1cJrQ0\\">
    <h3 class=\\"_1cJrQ0\\">level three</h3>
    <section class=\\"_1cJrQ0\\">
      <h4 class=\\"_1cJrQ0\\">level four</h4>
      <section class=\\"_1cJrQ0\\">
        <h5 class=\\"_1cJrQ0\\">level five</h5>
        <section class=\\"_1cJrQ0\\">
          <h6 class=\\"_1cJrQ0\\">level six</h6>
          <section class=\\"_1cJrQ0\\">
            <h6 class=\\"_1cJrQ0\\">level six again</h6>
            <h6 class=\\"_1cJrQ0\\">level six again</h6>
          </section>
          <h6 class=\\"_1cJrQ0\\">level six</h6>
        </section>
        <h5 class=\\"_1cJrQ0\\">level five</h5>
      </section>
      <h4 class=\\"_1cJrQ0\\">level four</h4>
    </section>
    <h3 class=\\"_1cJrQ0\\">level three</h3>
  </section>
  <h2 class=\\"_1cJrQ0\\">level two</h2>
</section>
"
`;
